File: bin/configs.go
package bin

import (
	"encoding/json"
	"net/http"
	"os"
	"path/filepath"
	"strings"

	"gopkg.in/yaml.v3"
)

func ConfigsHandler(w http.ResponseWriter, r *http.Request) {
	DebugLog("ConfigsHandler: method=%s", r.Method)
	switch r.Method {
	case http.MethodGet:
		getConfigs(w, r)
	case http.MethodPost:
		postConfigs(w, r)
	default:
		DebugLog("ConfigsHandler: method not allowed: %s", r.Method)
		http.Error(w, "405 Method Not Allowed", http.StatusMethodNotAllowed)
	}
}

func getConfigs(w http.ResponseWriter, r *http.Request) {
	ClientMutex.RLock()
	c := CurrentCore
	ClientMutex.RUnlock()

	if ct, ok := clientTypes[r.URL.Query().Get("core")]; ok { c = ct }
	DebugLog("getConfigs: using client=%s", c.Name)

	ext := c.ConfigExt
	if c.IsJSON { ext = "*.json" }

	patterns := []string{
		filepath.Join(c.ConfigDir, ext),
		filepath.Join(XkeenConf, "*.lst"),
	}

	var res []Config
	for _, p := range patterns {
		m, _ := filepath.Glob(p)
		for _, f := range m {
			if d, err := os.ReadFile(f); err == nil {
				res = append(res, Config{
					Name:     strings.TrimSuffix(filepath.Base(f), filepath.Ext(f)),
					Filename: filepath.Base(f),
					Content:  string(d),
				})
			}
		}
	}
	DebugLog("getConfigs: found %d configs", len(res))
	jsonResponse(w, ConfigsResponse{Success: true, Configs: res}, 200)
}

func postConfigs(w http.ResponseWriter, r *http.Request) {
	var req ActionRequest
	if json.NewDecoder(r.Body).Decode(&req) != nil {
		http.Error(w, "JSON error", 400)
		return
	}

	DebugLog("postConfigs: saving file=%s (%d bytes)", req.Filename, len(req.Content))

	ClientMutex.RLock()
	c := CurrentCore
	ClientMutex.RUnlock()

	name := filepath.Base(req.Filename)
	isLst := strings.HasSuffix(name, ".lst")
	path := filepath.Join(c.ConfigDir, name)

	if isLst {
		path = filepath.Join(XkeenConf, name)
		req.Content = strings.ReplaceAll(req.Content, "\r\n", "\n")
	} else {
		if c.IsJSON {
			if !strings.HasSuffix(path, ".json") { path += ".json" }
		} else {
			var y any
			if yaml.Unmarshal([]byte(req.Content), &y) != nil {
				http.Error(w, "Invalid YAML", 400)
				return
			}
		}
	}

	if req.Action == "save" {
		if os.WriteFile(path, []byte(req.Content), 0644) != nil {
			http.Error(w, "Write error", 500)
			return
		}
	} else {
		http.Error(w, "Unknown action", 400)
		return
	}
	jsonResponse(w, Response{Success: true}, 200)
}

File: src/configs.rs
use axum::{extract::{Query, State}, response::{IntoResponse, Json}};
use serde::{Deserialize, Serialize};
use std::{collections::HashMap, fs, path::PathBuf};
use crate::types::*;

#[derive(Serialize)]
struct ConfigItem { name: String, filename: String, content: String }

#[derive(Deserialize)]
pub struct ConfigReq { action: String, filename: String, content: String }

pub async fn get_configs(State(state): State<AppState>, Query(params): Query<HashMap<String, String>>) -> impl IntoResponse {
    let core = state.core.read().unwrap();
    let target = params.get("core").map(|s| s.as_str()).unwrap_or(&core.name);
    let (cdir, ext) = if target == "mihomo" { (MIHOMO_CONF, "config.yaml") } else { (XRAY_CONF, "json") };

    let mut res = Vec::new();

    if let Ok(entries) = fs::read_dir(cdir) {
        for entry in entries.filter_map(|e| e.ok()) {
            let path = entry.path();
            let matches = if target == "xray" {
                path.extension().map_or(false, |x| x == ext)
            } else {
                entry.file_name().to_str().map_or(false, |n| n == ext)
            };
            if matches {
                if let Ok(c) = fs::read_to_string(&path) {
                    res.push(ConfigItem {
                        name: path.file_stem().unwrap().to_string_lossy().into(),
                        filename: entry.file_name().to_string_lossy().into(),
                        content: c
                    });
                }
            }
        }
    }

    if let Ok(entries) = fs::read_dir(XKEEN_CONF) {
        for entry in entries.filter_map(|e| e.ok()) {
            let path = entry.path();
            if path.extension().map_or(false, |x| x == "lst") {
                if let Ok(c) = fs::read_to_string(&path) {
                    res.push(ConfigItem {
                        name: path.file_stem().unwrap().to_string_lossy().into(),
                        filename: entry.file_name().to_string_lossy().into(),
                        content: c
                    });
                }
            }
        }
    }

    res.sort_by(|a, b| a.name.cmp(&b.name));
    Json(serde_json::json!({ "success": true, "configs": res }))
}

pub async fn post_configs(State(state): State<AppState>, Json(req): Json<ConfigReq>) -> impl IntoResponse {
    let core = state.core.read().unwrap();
    let is_lst = req.filename.ends_with(".lst");
    let mut path = PathBuf::from(if is_lst { XKEEN_CONF } else { &core.conf_dir });
    path.push(&req.filename);

    let content = if is_lst { req.content.replace("\r\n", "\n") } else { req.content };

    if !is_lst && core.is_json && !path.to_string_lossy().ends_with(".json") {
        path.set_extension("json");
    }

    if req.action == "save" {
        if fs::write(path, content).is_err() {
            return Json(ApiResponse::<()> { success: false, error: Some("Write error".into()), data: None });
        }
    }

    Json(ApiResponse::<()> { success: true, error: None, data: None })
}